### DISCUSSION TITLE
Pixel / Android 14 Install Issues

### DISCUSSION THREAD

**Discussion by agronick** [agronick/aa-torque#23] [Q&A]
Making this so people can answer questions. I can't provide much help since I don't have one of these phones.

What to try:
- Follow the Kingstaller instructions for Pixel which involve reloading the package installer
- Root
- Root w/custom launcher

**Comment by ngknick** [comment]
Thanks for making this thread/ consolidating the pixel install issues. Agreed it's not a bug and appears device/ software specific. Good news is I was able to get it working just a few minutes ago. What seemed to rectify it for me was installing the installer package update 13, reinstalling King installer. Rebooting device. Then installing aa torque apk. 
![PXL_20231006_164859382 MP](https://github.com/agronick/aa-torque/assets/40443458/410d8f9f-8004-4a83-82bb-d1ed7a8d3fdc)

Pixel 7 pro on Google Fi (stock android)
Edit: added pic

**Reply by aross6** [reply]
Are you on Android 14 or 13? What is the exact name of the installerpackage you installed? Also, are you rooted?

**Reply by khalima** [reply]
@ngknick Thanks so much for your effort of trying to solve this issue. I was having the same issue with Pixel 5 and I managed to fix it with following your path.

Here's what I did: 
- I uninstalled King installer and AA-torque which I previously installed
- I installed/updated package installer 13 (https://www.apkmirror.com/apk/google-inc/package-installer/package-installer-13-10337213-release/)
- I installed King installer (https://github.com/fcaronte/KingInstaller#installation)
- I rebooted my phone (this is important, afaik)
- I installed AA Torque via King installer (https://github.com/agronick/aa-torque/releases)
..and it works. 🎉

**Reply by ryanmcdonnell** [reply]
This also worked for me on a Pixel 7 on Google Fi running Android 13.

**Reply by ngknick** [reply]
The question is how long until we are forced to update to 14? Hopefully they'll find a patch for it by then

**Reply by agronick** [reply]
People say AAAD works on Android 14. I took their code and used it for the in app update. So the new installation method without Kingstaller should work on Android 14. I'm curious to see what experiences people have with that.

**Reply by jamesnorman93** [reply]
I'm not having any success with the force update method without Kingstaller with my Pixel 6 on Android 14...

**Reply by Villacchione** [reply]
Can you share the process you followed to use the same  installing method of AAAD to update the app on Android 14 without using Kingstaller?

Thanks

**Comment by aross6** [comment]
> @ngknick Thanks so much for your effort of trying to solve this issue. I was having the same issue with Pixel 5 and I managed to fix it with following your path.
> 
> Here's what I did:
> 
> * I uninstalled King installer and AA-torque which I previously installed
> * I installed/updated package installer 13 (https://www.apkmirror.com/apk/google-inc/package-installer/package-installer-13-10337213-release/)
> * I installed King installer (https://github.com/fcaronte/KingInstaller#installation)
> * I rebooted my phone (this is important, afaik)
> * I installed AA Torque via King installer (https://github.com/agronick/aa-torque/releases)
>   ..and it works. 🎉

You must be on Android 13 because that version of PackageInstaller is not compatible with Android 14

**Reply by ngknick** [reply]
Correct my pixel 7 pro is on Android 13.

**Comment by Jonesyj83** [comment]
Pixel 6 pro on Android 14, Tried all the steps above, even tried Package Installer 14. Can't get it to work, Can get another app to work though for obd, Just has less features.

**Comment by Vince56190** [comment]
Correction : j'ai mis a jour l' application, elle ne fonctionne plus , n'apparait plus dans le menu Android Auto. Android 14 / Android Auto derniére version / AA Torque RC6.2. lors de la mise a jour sur Android 14 l'application fonctionnais encore, mais lorsque j'ai mis a jour l' app AA torque, elle a disparu de mon android Auto. je suppose donc que l' app fonctionne sur android 14 mais que king installer ne sait pas l'installer sur Android 14 .

Correction : I updated the application, it no longer works, no longer appears in the Android Auto menu. Android 14 / Android Auto latest version / AA Torque RC6.2. When updating to Android 14 the application was still working, but when I updated the AA torque app, it disappeared from my Android Auto. I therefore assume that the app works on Android 14 but that King Installer does not know how to install it on Android 14.

**Reply by agronick** [reply]
Did you try the new in app update method? It uses the same process AAAD uses.

**Reply by Vince56190** [reply]
I turned on my pixel 6a Android 14, I tested auto installation and kinginstaller, the application does not appear in the list of Android AUto applications

**Reply by agronick** [reply]
Does AAAD work when trying to install performance monitor?

**Reply by agronick** [reply]
I was looking at the AAAD bug reports. It sounds like people get it working if they update the package installer before they upgrade to Android 14. You can also downgrade, install the package installer, then upgrade. Can you show the error you get when you try to install the package installer on 14?

**Reply by Vince56190** [reply]
performance monitor installation still works with AAAD   . AA Torque made no installation error , it just dont appear on android Auto App list

**Reply by aross6** [reply]
Running a Pixel 6a with Android 14, Performance Monitor installed through AAAD shows up in the Android Auto apps list.

Unfortunately, AA Torque installed on Pixel 6a with Android 14 does not. The only way I've found to get AA Torque to show up in the Android Auto apps list is to downgrade to Android 13, install the google package installer app apk, install AA Torque using KingInstaller, then upgrade the phone to Android 14 (Note, updating the AA Torque after this process makes it disappear from the Android Auto apps list, so you're pretty much stuck with the AA Torque version you install).

Not sure why Performance Monitor installed through AAAD on Android 14 shows up in the Android Auto apps list but AA Torque installed while on Android 14 does not, but there must be some coding difference somewhere.

Irregardless, even after downgrading to Android 13, installing the google package installer app apk, installing AA Torque using KingInstaller, and the  upgrade the phone to Android 14, the AA Torque app crashes when trying to launch it from Android Auto. Hoping this gets resolved soon because it's a pain downgrading my phone, which wipes data, and having to reinstall everything just to try a new AA Torque release.

**Reply by agronick** [reply]
I think I can fix the crash. The only reason I can think of performance monitor working is because it targeted old SDKs. The code to install is the same between AAAD and this app: https://github.com/shmykelsa/AAAD/blob/main/app/src/main/java/com/legs/appsforaa/MainActivity.java#L1458 https://github.com/agronick/aa-torque/blob/master/app/src/main/java/com/aatorque/prefs/SettingsActivity.kt#L64

**Reply by agronick** [reply]
I thought of another reason. Performance monitor was debuggable on the production build. I changed the production build to be debuggable. I also added code which should fix the `SecurityException` on Android 14. Let me know if it works with the prerelease version either by bringing it to Android 14 or installing it directly.

**Reply by aross6** [reply]
Installer the new Prerelease directly on Android 14 either using the Package Installer or KingInstaller still does not yield it showing in the Android Auto apps list. On Android 14 I also installed it using KingInstaller and tried to do the Force Update in AA Torque, but it will not install as it returns the error "App not installed as package appears to be invalid".

I have not yet tried downgrading to A13, etc and installing the new Prerelease yet as I'm at work and it will take some time as this process requires wiping the phone and then reinstalling everything as it's my daily driver. Hopefully I can try it out this way soon though.

**Reply by aross6** [reply]
I can now confirm that AA Torque WORKS with Android 14 on the Google Pixel 6a by doing the following: Downgrading to Android 13, installing the Android 13 Package Installer apk, installing AA Torque 15.1 (10.0.87) using KingInstaller, and then updating to Android 14. So happy this version maintains across the A13->A14 update and the crash on launch on A14 is fixed. The only bummer is that because of the install restrictions, I'm stuck on AA Torque version 15.1 (10.0.87) and can't update since it will remove the app from the Android Auto apps list, but 15.1 (10.0.87) works great so I'm ok with that. Thanks so much for all your hard work!

**Reply by agronick** [reply]
That's great. If someone could figure out how to install a downgraded version of package installer I think you would be good for all versions.

**Reply by aross6** [reply]
Agreed!

**Reply by papatsonis** [reply]
I can also state the following steps that on my pixel 7(android 14) installed the "force update"version. Installed the latest with king installer, open/give permissions, and that time installed correctly from the aa-torque  the update. So seems like in the photo. Still need to return home to check actuall.. ![Screenshot_20231109-165829.png](https://github.com/agronick/aa-torque/assets/1456007/8ea871d8-8dee-4dac-abf0-3021ed95784e)

**Comment by Vince56190** [comment]
I tested the old version (performance Monitor) https://github.com/jilleb/mqb-pm with its installer https://github.com/shmykelsa/AAAD The application appears in Android Auto. Perhaps we need to check with the creator of AAAD, to put AATorque in place of Performance Monitor

**Reply by agronick** [reply]
THe new in app update code is the same thing AAAD is using. It should work now if AAAD works.

**Reply by Vince56190** [reply]
I turned on my pixel 6a Android 14, I tested auto installation and kinginstaller, the application does not appear in the list of Android AUto applications

**Comment by Vince56190** [comment]
AAAD  aswer : 
Thank you for this feedback.
We'll reach out the author and see if this can be added to AAAD. Regards

**Reply by jamesnorman93** [reply]
Looking forward to this outcome... I only just got AATorque and miss it greatly already... I guess it's too difficult to just add to the play store? 😭

**Reply by agronick** [reply]
They would never approve it. The previous developer tried. It uses a hacked SDK that they only give to OEMs.

**Reply by jamesnorman93** [reply]
Oh rats...

**Reply by aross6** [reply]
Have you tried submitting it through Google IOT (Internet of Things)? This is how the developer of DashOBD got his app in the Play Store. Here's the link: https://android-developers.googleblog.com/2023/04/expanding-options-for-building-apps-for-cars.html?m=1
There's also a thread about it on Reddit from that apps developer. May be worth a try?

**Reply by agronick** [reply]
They make you use templates. That is why the information on that app is so basic. You can only display a list or a grid of rows. If someone else finds a way around it I'd implement it. I played around with it and didn't see a way to get around it.

**Reply by aross6** [reply]
Could you make a "plugin" option and then utilize a downloaded gauge plugin to get to the desired interface?

**Reply by agronick** [reply]
I don't think so. I think there was a way to build your own templates by extending the template class but they patched it.

**Comment by ngknick** [comment]
Kyle which package installer? I'm on Android 13 everything with aa torque
still functional and willing to test.

sent from Project Fi

**Reply by agronick** [reply]
I think it is com.google.android.packageinstaller I'd be interested in seeing what happens when you install the 13 one on 14.

**Reply by aross6** [reply]
You can't install the Android 13 package installer on Android 14, because the Android 13 package installer uses SDK 33 and Android 14 requires SDK 34. Installing an Android 14 package installer apk does not result in AA Torque showing up in the Android Auto apps list either unfortunately.

**Reply by jamesnorman93** [reply]
Yep. This. 👌🏻
I tried the 13 package and it won't install.

**Comment by ngknick** [comment]
I see, and just to make sure I understand correctly:
1. Update to Android 14
2.Rollback/install package installer 13

3?Then reinstall kingstaller?
4? Install aa-torque using king?

sent from Project Fi

**Reply by agronick** [reply]
AAAD seems to work better than Kingstaller and the in app update method is the same as AAAD. I think you can only install package installer 13 on Android 13 but it doesn't get replaced when you upgrade. I need to see how it fails. But from what I saw on the AAAD issues page people had it working if they installed the non-Pixel copy of packageinstaller before they upgraded.

**Reply by jamesnorman93** [reply]
Correct - As soon as you try installing the aa-torque in Android 14, it disappears from the Android Auto list.

**Reply by jamesnorman93** [reply]
@agronick 
Is there a way I can report/log how it fails for you?

**Reply by agronick** [reply]
I tried getting it working in an emulator. It fails with `INSTALL_FAILED_VERSION_DOWNGRADE`. There are two ways I think this could work:

- Update the manifest of the older version to have the version code of the newer version
- Find a way to force a downgrade

The first one is hard because the manifest is iin a binary format and you would need a signing key to change it. There are flags to force a downgrade but they only work if the installed application is debuggable. What makes it even more hard is you are uninstalling the package installer which is the thing that installs package. You can uninstall it for user 0 with `adb shell pm uninstall --user 0 com.google.android.packageinstaller` but it doesn't install it from the system and it still breaks things.

**Reply by agronick** [reply]
Ii think there is a third option which is flash a zip upgrade in recovery mode but I'm not sure how to get it to pass the signing check.

**Comment by ngknick** [comment]
Okay, I think I follow now. Install the 14 package installer before
upgrading to 14, correct?

Then use aaad to install aa-torque?

sent from Project Fi

**Reply by jamesnorman93** [reply]
I don't think the 14 package will work on Android 13...?

Also, AAAD is a locked app with the developer choosing which apps are able to be chosen from. (It's not like the Kingstaller app where you can choose a custom package)
Apparently it's very unlikely the Developer would add AA-Torque to his app...? Which is crazy...

**Reply by agronick** [reply]
No, install the 13 package that isn't for Pixel. I'm not sure if it will work. Its just what people are saying here: https://github.com/shmykelsa/AAAD/issues/261

This is the comment I'm referring to:

> It works if u had android 13, ran the 13 package installer, & used it BEFORE going to android 14. If u upped to android 14 or ur phone came w/ 14 & u haven't already had it working ur screwed. All p8 owners are screwed.

**Comment by agronick** [comment]
I think I might have found a way to make it work on Android 14. I got it to show up in an emulator as downloaded from the play store. The latest prerelease is marked as debuggable. You will need to use adb. Put it in debug mode and connect it to your computer. Download the latest release from the releases page. Copy it to the device with `adb push`:

```
adb push DOWNLOAD_LOCATION  /data/local/tmp
```

You can then use the command line to install and set where it was installed from
```
adb shell  pm install -i "com.android.vending" /data/local/tmp/aa-torque.apk
```

You might have to do an uninstall first. Let me know if this works.

**Reply by aross6** [reply]
I will try this as soon as I get home. Just making sure, you're referring to Prerelease 15 version 10.0.86, correct?

**Reply by agronick** [reply]
15 or 15.1

**Reply by aross6** [reply]
Just tried pushing AA Torque (10.0.87) via ADB. The app shows as "App installed from Google Play Store", but unfortunately still does not show up in the Android Auto apps list.

**Reply by agronick** [reply]
You got developer mode on in Android auto?

**Reply by aross6** [reply]
Yes

**Reply by Strelica12** [reply]
Some feedback from my side, although not on Pixel phone.

I experienced same / similar behaviour as @aross6 (AA Torque 10.0.87 not showing up in AA apps list).

One thing i noticed when installing via ADB, is that AA Torque installed 2 apps (AA Torque and AA Torque Settings ????). First one i can open and it work, but not showing in AA apps list. Second one i can open but cannot do anything with it (i.e. cannot open any of the settings, not even info button).

Galaxy S23+, Android 14.

- Uninstalled AA Torque
- Clean AA cache
- Debug mode on phone enabled
- Pushed AA Torque (10.0.87) via ADB (USB method)
- AA Torque shows as "App installed from Google Play Store" in app info
- AA Torque listed in AA "Version and permission info"
- AA Torque Not Showing up in AA apps list

**Reply by agronick** [reply]
It sounds like the second one is the one that is supposed to launch in Android Auto. I'll need Samsung to put out a 14 update for the Note 20 ultra before I can try to get it working.

**Reply by spektr79** [reply]
I have the same phone as @Strelica12 , and of course the same symptoms. I am looking forward to enjoy using this app once again.

**Reply by DeltaTangoLima** [reply]
This method didn't work on my P7P with Android 14. App is installed, and says it was installed by Play Store, but still doesn't appear in the list of apps in AA. AA is definitely configured in Developer Mode, with Unknown Sources enabled.

Happy to assist with any testing of alternate solutions - P7P running vanilla Android 14, no root.

**Comment by J-Henvey** [comment]
> It sounds like the second one is the one that is supposed to launch in Android Auto. I'll need Samsung to put out a 14 update for the Note 20 ultra before I can try to get it working.

I tried with my Samsung Galaxy S23 Ultra running Android 14 and ran into no issues during the installation process but the app doesn't show up on Android Auto.

I am available if you need someone to test with on Android 14 because I would love to get it working! I just downloaded AAAD and will try to get it working that way.

If anyone any suggest please let me know!

**Comment by bornaradusin** [comment]
Pixel 7 with android 14 stock with magisk
1. Cleared cache and data from android auto 
2. Installed aa torque
3. Used aa aio tweaker to apply the patch so it looks like a app from play store
4. Rebooted the phone
5. Started the forced update in aa torque
6. Opened android auto setting and changed it to developer mode and enabled unknown sources 
7. Rebooted again (not sure if it was needed)
8. Tested and working fine

I've noticed sometimes it force closes but by the logs i figured out it doesn't open torque pro everytime and when it doesn't it starts force closing so using tasker to open torque pro when the bluetooth connects to my car to resolve this part.

Hopefully this helps out someone to get it running

**Reply by ronatron42** [reply]
This would require the phone be rooted no?

**Reply by bornaradusin** [reply]
Unfortunately yes ,AIO Tweaker requires root, i'm not sure if you could maybe root it, do the process and then remove magisk again that might be a viable solution but this is only if you have your bootloader unlocked otherwise you would need to unlock it firstly which would cause a complete formatting

**Comment by hogass83** [comment]
Had same issues after updating flip 5 to 14. Only one version of aa worked with 14. Older versions of aa would not work. Aa torque showed once then disappeared. Downgraded to 13. Downgraded installer to 13. Still would not show in aa. Went back to older aa version and reinstalled aa torque and boom it came back. I believe issue for me was in the newest version of android auto itself not android 14.

**Reply by Villacchione** [reply]
Which old version of Andoid Auto did you install? Thanks

**Reply by hogass83** [reply]
10.9.134603 beta from apkmirror with android 13

**Reply by xtonax** [reply]
Con android 13 però con android 14 non funziona

**Comment by aliatx2017** [comment]
Pixel7 w/Android14, not rooted. Ford Sync3.4 (wired AA)

Extracted and reinstalled packageinstaller14 (with X-plore file mgr), tried aa-torque without Kinginstaller first, force update, reboot. Not showing the app in AA. 
Then tried with Kinginstaller (uninstalled aa-torque, installed Kinginstaller, reboot, reinstalled aa-torque, etc.). Not showing the app in AA. 

Tried using AAWireless adapter, app not showing --> enabled developer mode --> SUCCESS
